tests/generic_config_updater/test_ntp.py

```diff
@@ -0,0 +1,228 @@
+import logging
+import pytest
+import re
+
+from tests.common.helpers.assertions import pytest_assert
+from tests.generic_config_updater.gu_utils import apply_patch, expect_op_failure, expect_op_success
+from tests.generic_config_updater.gu_utils import generate_tmpfile, delete_tmpfile
+from tests.generic_config_updater.gu_utils import create_checkpoint, delete_checkpoint, rollback_or_reload
+
+logger = logging.getLogger(__name__)
+
+pytestmark = [
+    pytest.mark.disable_loganalyzer,
+    pytest.mark.topology('any'),
+    pytest.mark.device_type('vs')
+]
+
+NTP_CONF         = "/etc/ntp.conf"
+NTP_SERVER_INIT  = "10.0.0.1"
+NTP_SERVER_DUMMY = "10.0.0.2"
+NTP_SERVER_RE    = "server {} iburst"
+
+@pytest.fixture(autouse=True)
+def setup_env(duthosts, rand_one_dut_hostname):
+    """
+    Setup/teardown fixture for ntp server config
+    Args:
+        duthosts: list of DUTs.
+        rand_selected_dut: The fixture returns a randomly selected DuT.
+    """
+    duthost = duthosts[rand_one_dut_hostname]
+    create_checkpoint(duthost)
+
+    init_ntp_servers = running_ntp_servers(duthost)
+
+    yield
+
+    try:
+        logger.info("Rolled back to original checkpoint")
+        rollback_or_reload(duthost)
+
+        cur_ntp_servers = running_ntp_servers(duthost)
+        pytest_assert(cur_ntp_servers == init_ntp_servers,
+            "ntp servers {} do not match {}.".format(cur_ntp_servers, init_ntp_servers)
+        )
+    finally:
+        delete_checkpoint(duthost)
+
+
+def running_ntp_servers(duthost):
+    """ Ger running ntp servers
+    """
+    config_facts = duthost.config_facts(host=duthost.hostname,
+                                        source="running")['ansible_facts']
+    ntp_servers = config_facts.get('NTP_SERVER', {})
+    return ntp_servers
+
+
+def ntp_server_test_setup(duthost):
+    """ Clean up ntp server before test
+    """
+    ntp_servers = running_ntp_servers(duthost)
+    for ntp_server in ntp_servers:
+        duthost.command("config ntp del %s" % ntp_server)
+
+
+def server_exist_in_conf(duthost, server_pattern):
+    """ Check if ntp server take effect in ntp.conf
+    """
+    content = duthost.command("cat {}".format(NTP_CONF))
+    for line in content['stdout_lines']:
+        if re.search(server_pattern, line):
+            return True
+    return False
+
+
+def ntp_service_restarted(duthost):
+    """ Check if ntp.service is just restarted within 10 secs
+    """
+    output = duthost.shell("systemctl show ntp.service --property ActiveState --value")
+    if output["stdout"] != "active":
+        return False
+    output = duthost.shell("ps -o etimes -p $(systemctl show ntp.service --property ExecMainPID --value) | sed '1d'")
+    if int(output['stdout'].strip()) < 10:
```

Comment:
I treat ntp service as restart when its starting time is less than 10sec.
Correct me if it is not reasonable.

Reply:
I am not sure how stable is the 10s hard-coded value. If you know exactly when the service is restarted (by your test script), you can take a timestamp before the restarting and  use it to check the `etimes`.

Reply:
The restart is embedded into GCU post-check. Seems hard to find the exact time.
10s should be enough. Usually, the elapsed time is just 1 sec.
I am a little worried that the service process is not up if the check is too soon.